lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -291,6 +293,46 @@ void SPIRVRegularizeLLVMBase::lowerUMulWithOverflow(
   UMulIntrinsic->setCalledFunction(UMulFunc);
 }
 
+void SPIRVRegularizeLLVMBase::adaptStructTypes(StructType *ST) {
+  if (!ST->hasName())
+    return;
+  StringRef STName = ST->getName();
+  STName.consume_front("struct.");
+  StringRef MangledName = STName.substr(0, STName.find('.'));
+
+  // Demangle the name of a template struct and parse the template
+  // parameters which look like:
+  // <signed char, 2ul, 2ul, (spv::MatrixLayout)0, (spv::Scope)3>
+  // The result should look like SPIR-V friendly LLVM IR:
+  // %spirv.JointMatrixINTEL._char_2_2_0_3
+  if (MangledName.startswith("_ZTSN5__spv24__spirv_JointMatrixINTEL")) {
+    std::string DemangledName = llvm::demangle(MangledName.str());
+    StringRef Name(DemangledName);
+    Name = Name.slice(Name.find('<') + 1, Name.rfind('>'));
+    std::stringstream SPVName;
+    SPVName << kSPIRVTypeName::PrefixAndDelim
+            << kSPIRVTypeName::JointMatrixINTEL << kSPIRVTypeName::Delimiter;
+    auto P = Name.split(", ");
+    StringRef ElemType = P.first;
+    if (size_t Space = ElemType.rfind(' '))
+      ElemType = ElemType.substr(Space + 1);
```

Comment:
What if there are no space? In which cases you expect the condition to return `false`?

Reply:
It is ok if there is no space. This `if` handles cases when we have type modifier, like in the example `signed char`. In this case we take only last word "char". If there is no space it means that there is no modifiers and `ElemType` already holds the name of the type we want.